nilearn/regions/hierarchical_kmeans_clustering.py

```diff
@@ -0,0 +1,276 @@
+        return array_round.astype(int)
+    elif np.sum(array_round) > n_clusters:
+        while np.sum(array_round) != n_clusters:
+            # prevent element rounded to 1 to be decreased in edge cases
+            mask = array_round != np.ones(array.size)
+            idx = np.argmin(array[mask] - array_round[mask])
+            parent_idx = np.arange(array_round.shape[0])[mask][idx]
+            array_round[parent_idx] -= 1
+        return array_round.astype(int)
+
+
+def hierarchical_k_means(X, n_clusters, init="k-means++", batch_size=1000,
+                         n_init=10, max_no_improvement=10, verbose=0,
+                         random_state=0):
+    """ Use a recursive k-means to cluster X. First clustering in sqrt(n_clusters)
+    parcels, and Kmeans a second time on each parcel. s
+
+    Parameters
+    ----------
+    X: ndarray (n_samples, n_features)
+        Data to cluster
+
+    n_clusters: int,
+        The number of clusters to find.
+
+    init : {'k-means++', 'random' or an ndarray}
+        Method for initialization, defaults to 'k-means++':
+        'k-means++' : selects initial cluster centers for k-means
+        clustering in a smart way to speed up convergence. See section
+        Notes in k_init for more details.
+        'random': choose k observations (rows) at random from data for
+        the initial centroids.
+        If an ndarray is passed, it should be of shape (n_clusters, n_features)
+        and gives the initial centers.
+
+    batch_size : int, optional, default: 1000
+        Size of the mini batches. (Kmeans performed through MiniBatchKMeans)
+
+    n_init : int, default=10
+        Number of random initializations that are tried.
+        In contrast to KMeans, the algorithm is only run once, using the
+        best of the ``n_init`` initializations as measured by inertia.
+
+    max_no_improvement : int, default: 10
+        Control early stopping based on the consecutive number of mini
+        batches that does not yield an improvement on the smoothed inertia.
+        To disable convergence detection based on inertia, set
+        max_no_improvement to None.
+
+    random_state : int, RandomState instance or None (default)
+        Determines random number generation for centroid initialization and
+        random reassignment. Use an int to make the randomness deterministic.
+
+    Returns
+    -------
+    labels : list of ints (len n_features)
+        Parcellation of features in clusters
+    """
+
+    n_big_clusters = int(np.sqrt(n_clusters))
+    mbk = MiniBatchKMeans(init=init, n_clusters=n_big_clusters,
+                          batch_size=batch_size, n_init=n_init,
+                          max_no_improvement=max_no_improvement,
+                          verbose=verbose, random_state=random_state).fit(X)
+    coarse_labels = mbk.labels_
+    fine_labels = np.zeros_like(coarse_labels)
+    q = 0
+    exact_clusters = np.asarray([n_clusters * np.sum(coarse_labels == i) * 1.
```

Comment:
you  could use `np.bincount` or `np.unique(..., return_counts=True)` to get the cluster sizes rather than repeatedly call `np.sum` and `==`